kubernetes/apps_kubeclient.go

```diff
@@ -770,63 +867,433 @@ func (kc *kubeClient) getPods(namespace string, uid types.UID) ([]v1.Pod, error)
+			for idx := range altBackends {
+				backend, ok := altBackends[idx].(map[interface{}]interface{})
+				if !ok {
+					return errors.New("Malformed alternative backend")
+				}
+				// Check if this alternate backend is a service we want a route for
+				backendKind, _ := backend["kind"].(string)
+				if backendKind == "Service" {
+					backendName, ok := backend["name"].(string)
+					if ok && len(backendName) > 0 {
+						_, pres := routesByService[backendName]
+						if pres {
+							matchingServices = append(matchingServices, backendName)
+						}
+					}
+				}
+			}
+		}
+		if len(matchingServices) > 0 {
+			// Get ingress points
+			status, ok := routeItem["status"].(map[interface{}]interface{})
+			if !ok {
+				return errors.New("Status missing from route")
+			}
+			ingresses, ok := status["ingress"].([]interface{})
+			if !ok {
+				return errors.New("No ingress array listed in route")
+			}
+
+			// Prefer ingress with oldest lastTransitionTime that is marked as admitted
+			oldestAdmittedIngress, err := findOldestAdmittedIngress(ingresses)
+			if err != nil {
+				return err
+			}
+
+			// Use hostname from oldest admitted ingress if possible
+			var hostname string
+			if oldestAdmittedIngress != nil {
+				hostname, ok = oldestAdmittedIngress["host"].(string)
+				if !ok {
+					return errors.New("Hostname missing from ingress")
+				}
+			} else {
+				// Fall back to optional host in spec
+				hostname, _ = spec["host"].(string)
+			}
+
+			// Check for optional path
+			path, _ := spec["path"].(string)
+
+			// Determine whether route uses TLS
+			// see: https://github.com/openshift/origin-web-console/blob/v3.7.0/app/scripts/filters/resources.js#L193
+			isTLS := false
+			tls, ok := spec["tls"].(map[interface{}]interface{})
+			if ok {
+				tlsTerm, ok := tls["termination"].(string)
+				if ok && len(tlsTerm) > 0 {
+					isTLS = true
+				}
+			}
+
+			// Check if this route uses a custom hostname
+			customHost := true
+			metadata, ok := routeItem["metadata"].(map[interface{}]interface{})
+			if ok {
+				annotations, ok := metadata["annotations"].(map[interface{}]interface{})
+				if ok {
+					hostGenerated, _ := annotations["openshift.io/host.generated"].(string)
+					if hostGenerated == "true" {
+						customHost = false
+					}
+				}
+			}
+			route := &route{
+				host:                 hostname,
+				path:                 path,
+				tls:                  isTLS,
+				hasAdmitted:          oldestAdmittedIngress != nil,
+				hasAlternateBackends: len(altBackends) > 0,
+				isCustomHost:         customHost,
+			}
+			// TODO check wildcard policy? (see above link)
+			// Associate this route with any services whoses routes we're looking for
+			for _, serviceName := range matchingServices {
+				routesByService[serviceName] = append(routesByService[serviceName], route)
+			}
+		}
+	}
+	return nil
+}
+
+func findOldestAdmittedIngress(ingresses []interface{}) (ingress map[interface{}]interface{}, err error) {
```

Comment:
is this method covered by 1 or more tests ?

Reply:
A whitebox test has not yet been written.